clang/lib/Driver/ToolChains/Clang.cpp

```diff
@@ -4117,21 +4119,24 @@ void Clang::ConstructJob(Compilation &C, const JobAction &JA,
                      options::OPT_fno_sycl_allow_func_ptr, false)) {
       CmdArgs.push_back("-fsycl-allow-func-ptr");
     }
-  }
 
-    if (IsSYCL) {
-    if (Arg *A = Args.getLastArg(options::OPT_sycl_std_EQ)) {
-      A->render(Args, CmdArgs);
-      CmdArgs.push_back("-fsycl-std-layout-kernel-params");
-    } else {
-      // Ensure the default version in SYCL mode is 1.2.1 (aka 2017)
-      CmdArgs.push_back("-sycl-std=2017");
+    if (UseSYCLTriple && !SYCLStdArg) {
       // The user had not pass SYCL version, thus we'll employ no-sycl-strict
       // to allow address-space unqualified pointers in function params/return
       // along with marking the same function with explicit SYCL_EXTERNAL
       CmdArgs.push_back("-Wno-sycl-strict");
     }
+  }
+
+  if (IsSYCL) {
+    if (SYCLStdArg) {
+      SYCLStdArg->render(Args, CmdArgs);
+      CmdArgs.push_back("-fsycl-std-layout-kernel-params");
+    } else {
+      // Ensure the default version in SYCL mode is 1.2.1 (aka 2017)
+      CmdArgs.push_back("-sycl-std=2017");
     }
+  }
```

Comment:
Would a further rearrangement achieve the purpose as well?
```suggestion
    if (Arg *SYCLStdArg = Args.getLastArg(options::OPT_sycl_std_EQ)) { 
      SYCLStdArg->render(Args, CmdArgs);
      CmdArgs.push_back("-fsycl-std-layout-kernel-params");
    } else {
      // The user had not pass SYCL version, thus we'll employ no-sycl-strict
      // to allow address-space unqualified pointers in function params/return
      // along with marking the same function with explicit SYCL_EXTERNAL
      CmdArgs.push_back("-Wno-sycl-strict");
      // Ensure the default version in SYCL mode is 1.2.1 (aka 2017)
      CmdArgs.push_back("-sycl-std=2017");
    }
  }
```
(and thus removing the pre-extraction at line 4085).

P.S. Sorry for iterating several times

Reply:
Not really sure that this rearrangement will fit fix this patch represents.

Reply:
I mean, with `-fsycl-device-only`, `UseSYCLTriple` should probably be set to `true`. Therefore, we'd cover the case, and any fathomable scenarios where a SYCL triple is used. Is this assumption flawed in some way?